ledger/participant-integration-api/src/main/scala/platform/store/backend/StorageBackend.scala

```diff
@@ -283,9 +287,13 @@ object DBLockStorageBackend {
 }
 
 object StorageBackend {
-  case class Params(ledgerEnd: Offset, eventSeqId: Long)
-
-  case class LedgerEnd(lastOffset: Option[Offset], lastEventSeqId: Option[Long])
+  case class LedgerEnd(lastOffset: Offset, lastEventSeqId: Long)
+  case class OptionalLedgerEnd(lastOffset: Option[Offset], lastEventSeqId: Option[Long])
```

Comment:
I thought about this, but I was worried about corner cases:
- The ledger_end offset moves for all kind of updates (including, e.g., party allocations), while the ledger_end_sequential_id only moves for `TransactionAccepted` updates.
- You can have an existing mutating schema database with a ledger end offset, but no events. If you migrate it, `ledger_end_sequential_id` will be `NULL`.

We could add a migration that fills the `ledger_end_sequential_id` column with 0 if the value is missing but `ledger_end` is defined. This would hardcode the constant for "no event yet" in a SQL migration, but that's ok. The the invariant you mentioned would hold.

Reply:
> The ledger_end offset moves for all kind of updates (including, e.g., party allocations), while the ledger_end_sequential_id only moves for TransactionAccepted updates.

That is true, but the ledger end never the less is set by setting them both. We do not set only one.

> You can have an existing mutating schema database with a ledger end offset, but no events. If you migrate it, ledger_end_sequential_id will be NULL.

That is true, but that can be circumvented by falling back to the default upon null, because this corner case allows it (offset some, event_seq_id null case). No additional migration is needed.

Reply:
I have refactored the code according to this discussion.